### PR TITLE
do not borrow non-Sync data in constants

### PR DESCRIPTION
**Pull request by RalfJung**
We cannot share that data across threads. non-Sync is as bad as non-Freeze in that regard.

This is currently WIP because it ignores a test that is broken by https://github.com/rust-lang/rust/issues/54419. But it is good enough to get crater going.

Fixes https://github.com/rust-lang/rust/issues/49206.

Cc @eddyb @nikomatsakis

### CODE REVIEW

File: src/test/ui/inhabitedness-infinite-loop.rs [hunk 1]
```diff
@@ -9,6 +9,7 @@
 // except according to those terms.
 
 // error-pattern:reached recursion limit
+// ignore-test FIXME #54419
```
Comment (RalfJung) [inline src/test/ui/inhabitedness-infinite-loop.rs]: This is the WIP part.

**Comment by RalfJung** [comment]
Oh great. Not even the full compiler bootstraps with this change.^^

**Comment by RalfJung** [comment]
Seems like `rustc::middle::resolve_lifetime::Scope` is not `Sync` but we have a const with a reference to it?

Why is it not `Sync`?

**Comment by RalfJung** [comment]
I hacked around this, so that we can get a crater run.

@bors try

**Comment by pietroalbini** [comment]
@craterbot run start=master#1002e404e1ae6b8b907c8655edd41380d0c561cb end=try#4056754e28b7351841378254fef9dbe2f2357c48 mode=check-only

**Comment by pietroalbini** [comment]
@craterbot abort

**Comment by pietroalbini** [comment]
@craterbot run start=master#1002e404e1ae6b8b907c8655edd41380d0c561cb end=try#4056754e28b7351841378254fef9dbe2f2357c48 mode=check-only

**Comment by pietroalbini** [comment]
Sorry, there was an error during the Crater redeploy, had to restart the experiment.

**Comment by RalfJung** [comment]
Also somehow `None` handling regressed, even for types that *are* `Sync`. Not sure how that happened. (That's the CI failure.)

**Comment by RalfJung** [comment]
I tried fixing the failing test but that is far from trivial.

Even with more refined HIR checking, we fail because `None` is used as constructor for the following two non-`Sync` types: `Option<Cell<String>` and `Option<Box<T>>` (where `T` is a type variable).

To justify why this could be accepted, we have to argue that `Option` is a type with a structural implementation of `Sync`, and that hence `None` cannot make things non-`Sync`. But this is particular to `Option`; another isomorphic type could do `impl<T> !Sync for MyOption<T>` and then we would have to reject even `MyOption::None`.

I could try hard-coding `core::option::Option::None` as a white-listed `Sync` value, but that's about the best we can do I think.

**Comment by RalfJung** [comment]
There may be something slightly better and still almost sound: We can accept a `VariantCtor` with no arguments if that type does NOT have an `impl !Sync`.

But I don't like it, because one can imagine a user doing
```rust
enum MyNonSyncOption<T> {
  NonSync(Cell<()>),
  Some(T),
  None
}
```
and then relying on the fact that this type is not `Sync`, using `MyNonSyncOption` as a token in whatever way.

So we could go one step further and instead check the following: If we have a constructor with no arguments (i.e., as a `hir::ExprKind::Path`), then we replace all type parameters by `()` and *then* we check `Sync`. I am not sure if that is even possible, but I cannot come up with a counterexample to break this. It is also horrible.

**Comment by eddyb** [comment]
@RalfJung Oh wow. One thing you can do is take the definition of the type (e.g. `Option<T>`), extend its `ParamEnv` to include `Sync` bounds for all parameters (i.e. `T: Sync`), and *check that* for `Sync`.
If a type passes that check, then I believe "value-based reasoning" can apply to it.

**Comment by RalfJung** [comment]
Yeah that sounds more reasonable than using `()` (which might violate other bounds).

**Comment by RalfJung** [comment]
Actually I am not so sure. This is still not filly justified.

First of all, we are assuming some kind of parametricity: That value would be `Sync` when considered at a different type (where the type parameters are all `Sync`), hence the same value will still be `Sync` even with non-`Sync` type parameters. That's dubious at best, the nature of a value can change dramatically as its type changes (consider that `T` and `Cell<T>` contain the same values).

Secondly, the type may not actually be instantiatable with non-`Sync` types. Imagine
```rust
trait Foo {} // private trait!

impl<T> Foo for Cell<T>

enum MyOption<T: Foo> {
  Some(T), None
}
```
I think `&MyOption::None` should NOT be allowed in a const. The author of this library can rely on the fact that `Foo` is only implemented for non-`Sync` types, and hence `MyOption` does not have to be `Sync`. ever. There is no reason to assume that `MyOption::None::<T>` should be a `Sync` value for any `T`, and I think it would be sound for unsafe code to assume it is not.
(Basically, imagine `Foo` had extra custom invariants attached to it that `MyOption` exploits. Technically it should be an `unsafe trait` then but traditionally we have not required `unsafe` on private items that a crate controls.)

**Comment by pietroalbini** [comment]
@craterbot retry-report

**Comment by RalfJung** [comment]
As expected, too many regressions. :/ Most of them involve `None` or things like `Sequence::Empty` and would probably be fixed by a more permissive value-based reasoning for `Sync`. I don't really like any of the proposals we had there but I feel we need something :(

However, there are others. [libdwfl](https://crater-reports.s3.amazonaws.com/pr-54424/try%234056754e28b7351841378254fef9dbe2f2357c48/reg/libdwfl-0.0.5/log.txt) has a `&ffi::Dwfl_Callbacks` constant where `ffi::Dwfl_Callbacks` contains a `*mut` which is not `Sync`; same for . [ffi_utils](https://crater-reports.s3.amazonaws.com/pr-54424/try%234056754e28b7351841378254fef9dbe2f2357c48/reg/ffi_utils-0.10.0/log.txt) has something similar with a `*const`. [jobserver](https://crater-reports.s3.amazonaws.com/pr-54424/try%234056754e28b7351841378254fef9dbe2f2357c48/reg/jobserver-0.1.11/log.txt) has `Fn()` in its contants, which needs to be `Fn() + Sync`. [rustfmt-nightly](https://crater-reports.s3.amazonaws.com/pr-54424/try%234056754e28b7351841378254fef9dbe2f2357c48/reg/rustfmt-nightly-0.99.4/log.txt) fails because `ast::Visibility` is not `Sync`. [rustcache-lists](https://crater-reports.s3.amazonaws.com/pr-54424/try%234056754e28b7351841378254fef9dbe2f2357c48/reg/rustache-lists-0.1.2/log.txt) and [tylar](https://crater-reports.s3.amazonaws.com/pr-54424/try%234056754e28b7351841378254fef9dbe2f2357c48/reg/tylar-0.2.2/log.txt) need the recursion limit bumped.

**Comment by RalfJung** [comment]
I should add that I think the jobserver case shows we need this. That's a trait object without a `Sync` bound, so this could contain *any kind of* non-`Sync` data and access it concurrently from multiple threads.

@eddyb Btw, do the value-based `Freeze` checks work correctly for trait objects...?

**Comment by RalfJung** [comment]
Okay, closures are actually fine because we only promote them if they do not capture anything -- so the real type is `()`, and by value-based reasoning this is `Freeze` and `Sync`.

**Comment by glaebhoerl** [comment]
(Given that this is a soundness fix and we'd de jure be within our rights to break things, might some kind of ugly hack to keep things compiling *for now* together with a deprecation/removal schedule for it (escalating lint levels, as usual) be a way forward?)

**Comment by RalfJung** [comment]
We should come up with a reasonable story then though for what crates should do instead. Clearly there is demand for value-based reasoning.

**Comment by eddyb** [comment]
This is *not* a soundness fix like any other. You have to write unsafe code that assumes something that I don't even think we explicitly guarantee *anywhere* already.
(That is, something about the address of a value of a type that doesn't implement `Sync`).

@RalfJung Did you run crater with *only* checking `Sync` in functions, *not* `const` / `static`?
I think that would paint a more accurate picture, as to the breakage of *promotion*.
Whereas for enforcing `Sync` in `const`, we might want to spend some time developing a properly auto-trait-integrated "value-based reasoning" system.

**Comment by RalfJung** [comment]
@eddyb 
> You have to write unsafe code that assumes something that I don't even think we explicitly guarantee anywhere already, to observe a problem.

We make so few guarantees that this argument applies way too easily.

> (That is, something about the address of a value of a type that doesn't implement Sync).

No, you do not have to observe the address. All you need to be affected by this problem is a `Freeze` non-`Sync` token type that represents the right of this thread to perform mutation elsewhere -- something that @pythonesque and others have been using e.g. for their "zero-overhead thread-safe `RefCell` with a key". Their concrete data types are likely not affected because the token types there contain lifetimes and I do not think generativity works in `const`. But a variant that replaces the lifetimes by run-time checks would be affected, I think.

Non-`Sync` references in `const` require (if a hypothetical proof of correctness) an assumption that is otherwise not made in the compiler. They can *not* be explained just in terms of syntactic sugar. This new assumption is a special exception, breaking a principle that otherwise holds true in general (non-`Sync` values cannot be shared across threads).

**Comment by eddyb** [comment]
You can't have unsafe code affected by this *from the outside*. That is, any sound abstraction could only be affected *from within* the abstraction barrier.

I'm not saying we shouldn't fix this, just that we can't be jumping to "soundness fix, break the world".

**Comment by RalfJung** [comment]
> any sound abstraction could only be affected from within the abstraction barrier.

I cannot see how this should be true. It is certainly not true on nightly for people using the `const_fn` feature.

**Comment by RalfJung** [comment]
Btw, the problem with `Scope` ultimately comes from `Span` not being `Sync`, which makes `hir::ParamName` non-`Sync`.

**Comment by RalfJung** [comment]
> Whereas for enforcing Sync in const, we might want to spend some time developing a properly auto-trait-integrated "value-based reasoning" system.

Both are unsound for the same reason, so what is the point in fixing one but not the other?

**Comment by RalfJung** [comment]
I implemented value-based reasoning as sketched above -- not because I am 100% sure it is sound, but because the overall result makes this *more sound* than it was before this PR. I do not expect to convince people to actually get rid of value-based reasoning for `Sync`, but I hope this is restricted enough that it will not lead to problems in practice.

@bors try

**Comment by RalfJung** [comment]
@rust-lang/infra could we get a crater run (for https://github.com/rust-lang/rust/commit/e927cbc923a535f55899e0567cb024b588343fbc)?

**Comment by kennytm** [comment]
@craterbot run start=master#9653f790333d1270f36f1614e85d8a7b54193e75 end=try#e927cbc923a535f55899e0567cb024b588343fbc mode=check-only

**Comment by kennytm** [comment]
@craterbot run start=master#9653f790333d1270f36f1614e85d8a7b54193e75 end=try#e927cbc923a535f55899e0567cb024b588343fbc mode=check-only name=pr-54424-2

**Comment by RalfJung** [comment]
That didn't go as well as I had hoped. :/

The most common failure pattern is something like:

Oct 03 16:37:12.991 INFO kablam! 135 |   pub const FFI_RESULT_OK: &FfiResult = &FfiResult {
Oct 03 16:37:12.991 INFO kablam!     |  _______________________________________^
Oct 03 16:37:12.992 INFO kablam! 136 | |     error_code: 0,
Oct 03 16:37:12.993 INFO kablam! 137 | |     description: 0 as *const c_char,
Oct 03 16:37:12.993 INFO kablam! 138 | | };
Oct 03 16:37:12.994 INFO kablam!     | |_^

Value-based, this seems fine. However, this is still constructing an instance of a non-Sync type, and there is no way to know that it's non-Sync for no good reason, at least for these values.

Basically, if I have a struct containing a non-Sync type, I can rely on it not being Sync. It doesn't matter that all values floating into the constructor are Sync, wrapping a newtype around them can change the game. This breaks ffi_utils, jobserver, libdwfl and is likely also the cause for yargb. Something similar can happen for enums (being inherently non-Sync because some variant is), and that breaks unison, sharing-list and sexpr-rust.

seq could be fixed by improving value-based reasoning for enums to ignore lifetime bounds.

rustfmt (indirectly) has a Span that it wants to share across thread in a const, which is not Sync.

**Comment by TimNN** [comment]
Ping from triage @RalfJung: What is the status of this PR?

**Comment by TimNN** [comment]
Ping from triage @rust-lang/lang: It looks like this PR could use your input.

If I understand things correctly, the current issue is that a specific pattern (`const`s containing references to other non-sync `const`s) would become an error with this PR, but is used by some crates. It is not immediately obvious that the this specific pattern is unsound except in some very contrived examples. 

This PR also has some other effects I believe, but I'm not sure what exactly they are.

@RalfJung: Please verify that my summary is correct / add anything I have missed. It would then probably make sense to `I-nominate` this PR for lang-team discussion.

File: src/librustc/ty/util.rs [hunk 2]
```diff
@@ -853,6 +861,50 @@ impl<'a, 'tcx> ty::TyS<'tcx> {
...
         r
     }
+
+    /// Returns a potentially different type that
+    /// can be used to check whether a 0-argument constructor of this type is
+    /// a `Sync` value.
+    /// Concretely, some type parameters get replaced by `()`.  We assume that
+    /// if a type parameter has no bounds (except for `Sized`), and if the same
+    /// type is `Sync` for that type parameter replaced by `()`, then the constructor
+    /// valie is also `Sync` at the original type.  This is a kind of parametricity
```
Comment (oli-obk) [inline src/librustc/ty/util.rs]: valie -> value

File: src/librustc/ty/util.rs [hunk 3]
```diff
@@ -853,6 +861,50 @@ impl<'a, 'tcx> ty::TyS<'tcx> {
...
+    /// assumption.
+    /// For now, we only do anything if *no* type parameter has any bound other than
```
Comment (oli-obk) [inline src/librustc/ty/util.rs]: anything -> something

File: src/librustc_mir/transform/qualify_consts.rs [hunk 4]
```diff
@@ -788,16 +791,36 @@ impl<'a, 'tcx> Visitor<'tcx> for Qualifier<'a, 'tcx, 'tcx> {
...
-                        self.add_type(ty);
-                        assert!(self.qualif.contains(Qualif::MUTABLE_INTERIOR));
+                    let ty = rvalue.ty(self.mir, self.tcx);
+                    // Value-based reasoning:
+                    // We are looking at a concrete type constructor, and we know
+                    // the only way to construct "fresh" non-Freeze data is `UnsafeCell`.
+                    // So we can check for that instead of `Freeze`.
+                    let freeze = Some(def.did) != self.tcx.lang_items().unsafe_cell_type();
+                    // For Sync, we generally have to fall back on the type, but have a
+                    // special exception for 0-argument ADT cosntructors.
```
Comment (oli-obk) [inline src/librustc_mir/transform/qualify_consts.rs]: cosntructors -> constructors

File: src/librustc_mir/transform/qualify_consts.rs [hunk 5]
```diff
@@ -788,16 +791,36 @@ impl<'a, 'tcx> Visitor<'tcx> for Qualifier<'a, 'tcx, 'tcx> {
...
+                    let sync = {
+                        let generalized_ty = if args.len() == 0 {
+                            // A type cosntructor with no arguments is directly a value.
```
Comment (oli-obk) [inline src/librustc_mir/transform/qualify_consts.rs]: cosntructor

**Review by oli-obk — COMMENTED** [review]
I understand that these are weird cases, but the fact that these things can be constructed at compile-time does have some meaning (yes I read the issue).

I think we could reduce the breaking change to runtime promoteds only and keep allowing `!Sync` but `Freeze` data in constants.

**Comment by eddyb** [comment]
@oli-obk Maybe we can emit lint warnings (from the MIR checking code) for `const`s?
Not necessarily saying we'll break it, but... I don't know.
I wish we had a proper model for value-based reasoning, because it's *so good*.

**Comment by RalfJung** [comment]
@TimNN: Basically, yes. There is no grounds on which we can allow non-`Sync` references in constants: We are compiling code with behavior that would be impossible to obtain without these references. This is strictly expanding the power of safe code. To make things worse, it is entirely unclear what the proof obligation looks like that unsafe code has to satisfy to justify this extra power that safe code has.

Yet, with the current limitations of `const`, the counter-examples I can construct are rather contrived and based on comparing pointer addresses. But I am very worried that we might extend the scope of `const` to the extend that this soundness hole becomes easier to exploit.

I will try to be in the lang team meeting today to discuss this.

**Comment by RalfJung** [comment]
Some potential hacks I could imagine:

* Only reject references to type that explicitly `impl !Sync`.
* When checking for `Sync` here, assume that raw pointers are `Sync`. (That is a sound assumption, and from what I recall it would fix most of the bad examples. However, I am not sure if this is feasible.)

**Comment by RalfJung** [comment]
Summary of lang team discussion: Nobody objected to this being a soundness issue *per se*, but for lack of actually exhibited UB we also shouldn't just break code. I am going to try and make this a warning instead of an error, redirecting people to a tracking issue that explains what this is about. I think for promotion we can likely make it a hard error now.

I also noticed that almost all of the remaining regressions are due to raw pointers. They are by far the most common `Freeze` but non-`Sync` type -- and they *could* be `Sync`; the reason they are not is essentially just a lint. We cannot change that now, but I proposed that we could do the `Sync` check for references in `const` under a hypothetical logical context where raw pointers *do* implement `Sync`. That wouldn't be entirely sound either, but it would likely be better than the current situation...

**Comment by Mark-Simulacrum** [comment]
>  hypothetical logical context where raw pointers do implement Sync. That wouldn't be entirely sound either, but it would likely be better than the current situation...

This seems unlikely to be ideal since unsafe code outside of the compiler/std is likely to use `PhantomData` with raw pointers to opt-out of sync intentionally, as negative impls are unstable.

--------

More generally, will you have a chance to update this PR soon? Or maybe lang team needs to come to a decision on it?

**Comment by RalfJung** [comment]
I hope to get to it next week.

**Comment by TimNN** [comment]
Ping from triage @RalfJung: What is the status of this PR?

**Comment by RalfJung** [comment]
Ah sorry, I didn't have time to get to this. :/  I still want to do this, but it could take a few more weeks.

**Comment by TimNN** [comment]
Ping from triage! If I understand recent comments correctly, this PR won't make any progress for a while (I really wish there was that time-machine 🙄), so I'm closing it as inactive for now. Thanks for your contributions, @RalfJung!

**Pull request closed without merging**